dnf5-plugins/system_upgrade_plugin/system_upgrade.cpp

```diff
@@ -0,0 +1,517 @@
+}
+
+void SystemUpgradeDownloadCommand::run() {
+    auto & ctx = get_context();
+
+    const auto & goal = std::make_unique<libdnf5::Goal>(ctx.base);
+
+    auto & conf = ctx.base.get_config();
+    conf.opt_binds().at("tsflags").new_string(libdnf5::Option::Priority::RUNTIME, "test");
+
+    if (no_downgrade->get_value()) {
+        const auto & settings = libdnf5::GoalJobSettings();
+        goal->add_rpm_upgrade(settings);
+    } else {
+        goal->add_rpm_distro_sync();
+    }
+
+    auto transaction = goal->resolve();
+    if (transaction.get_problems() != libdnf5::GoalProblem::NO_PROBLEM) {
+        throw libdnf5::cli::GoalResolveError(transaction);
+    }
+
+    ctx.download_and_run(transaction);
+
+    if (!transaction.get_transaction_packages_count()) {
+        throw libdnf5::cli::CommandExitError(
+            1, M_("The system-upgrade transaction is empty; your system is already up-to-date."));
+    }
+
+    const std::string json = transaction.serialize();
+    const auto & transaction_json_path = std::filesystem::path(get_datadir()) / "system-upgrade-transaction.json";
+    auto transaction_json_file = libdnf5::utils::fs::File(transaction_json_path, "w");
+
+    transaction_json_file.write(json);
+    transaction_json_file.close();
+
+    auto state = get_state();
+    state.get_data().cachedir = get_cachedir()->get_value();
+    state.get_data().system_releasever = system_releasever;
+    state.get_data().target_releasever = target_releasever;
+    state.get_data().status = STATUS_DOWNLOAD_COMPLETE;
+    libdnf5::repo::RepoQuery enabled_repos(ctx.base);
+    enabled_repos.filter_enabled(true);
+    enabled_repos.filter_type(libdnf5::repo::Repo::Type::SYSTEM, libdnf5::sack::QueryCmp::NEQ);
+    for (const auto & repo : enabled_repos) {
+        state.get_data().enabled_repos.emplace_back(repo->get_id());
+    }
+    libdnf5::repo::RepoQuery disabled_repos(ctx.base);
+    disabled_repos.filter_enabled(false);
+    disabled_repos.filter_type(libdnf5::repo::Repo::Type::SYSTEM, libdnf5::sack::QueryCmp::NEQ);
+    for (const auto & repo : disabled_repos) {
+        state.get_data().disabled_repos.emplace_back(repo->get_id());
+    }
+    state.write();
+
+    std::cout << "Download complete! Use `dnf5 system-upgrade reboot` to start the upgrade." << std::endl
+              << "To remove cached metadata and transaction, use `dnf5 system-upgrade clean`." << std::endl;
+}
+
+void check_state(const OfflineTransactionState & state) {
+    const auto & read_exception = state.get_read_exception();
+    if (read_exception != nullptr) {
+        try {
+            std::rethrow_exception(read_exception);
+        } catch (const std::exception & ex) {
+            const std::string message{ex.what()};
+            throw libdnf5::cli::CommandExitError(
+                1, M_("Error reading state: {}. Rerun `dnf5 system-upgrade download [OPTIONS]`."), message);
+        }
+    }
+}
+
+void reboot(Context & ctx, bool poweroff = false) {
```

Comment:
I recall a request to include a user confirmation, along with an explanation of the upcoming actions, to prevent accidental reboots and upgrades. However, I am unable to find the ticket at the moment.

Reply:
Resolved in #1280, and `sudo dnf5 offline reboot -y` will skip the confirmation.